**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Joshua Gamboa Calvo
*   MATRÍCULA: A01841021


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [82]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [83]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [84]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

In [85]:
# Creacion de funcion reutilizable para el llamado a base de datos basado en un SQL query

def query_db(query_str: str):
    query = sqla.text(query_str)
    return pd.read_sql(query, conn)

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [86]:
query_db("""
  SELECT *
  FROM movie
  WHERE runtime > 300
""")

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [87]:
# Explicacion: Se toma la columna de available globally para indentificar si ha
# sido publicado fuera de estados unidos, antes de eso se filtra con el where clause.
# para calcular el porcentaje solo se tome el count y se utiliza un flag para
# sumar todas las ocurrencias donde available globally es false.

query_db("""
  SELECT
    ROUND(
      100.0 * SUM(CASE WHEN available_globally = FALSE THEN 1 ELSE 0 END) / COUNT(*),
      2
    ) AS only_eu
  FROM movie
  WHERE available_globally IS NOT NULL
""")

,only_eu
0,83.57


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [88]:
query_db("""
  SELECT
    locale, count(*) as total
  FROM movie
  GROUP BY locale
""")

,locale,total
0,None,11255
1,en,532


### ¿Cuántos registros tienen el campo locale con valor NULL? (NULL en SQL ⇔ None en Python)

Basado en el query realizado en la celda anterior, tenemos que solo existen peliculas donde el idioma original es ingles `en` donde en total son `532`, y la cuenta de nulos en el campo `locale` son `11255`.

Muy probablemente, basado en esta información y en análisis previos, esto ocurre debido a que cuando la pelicula no es en ingles, no se registra su idioma en la columna locale.

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [89]:
# Contar la cantidad de ocurrencias donde los titulos originales y el titulo no
# coinciden

query_db("""
  SELECT
    count(*) non_matching_titles_count
  FROM movie
  WHERE title <> original_title
  AND locale is NULL
""")

,non_matching_titles_count
0,3947


In [90]:
query_db("""
  SELECT *
  FROM movie
  WHERE title = original_title
  AND locale is NULL
""")

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,1089,2024-01-01,2024-01-01,b'\x00',None,Recep İvedik 7,None,133,Recep Ivedik 7
1,1349,2024-01-01,2024-01-01,b'\x00',None,Verónica,None,105,Veronica
2,3509,2024-01-01,2024-01-01,b'\x00',None,Fantômas,None,100,Fantomas
3,3611,2024-01-01,2024-01-01,b'\x00',None,Luccas Neto em: O Hotel Mágico,None,70,Luccas Neto em: O Hotel Magico
4,4566,2024-01-01,2024-01-01,b'\x00',None,Un día en el paraíso,None,97,Un Dia En El Paraiso
5,5077,2024-01-01,2024-01-01,b'\x01',None,Superlópez,None,108,Superlopez
6,5211,2024-01-01,2024-01-01,b'\x00',None,Barış Akarsu Merhaba,None,115,Baris Akarsu Merhaba
7,5684,2024-01-01,2024-01-01,b'\x00',None,Tótem,None,95,Totem
8,6427,2024-01-01,2024-01-01,b'\x01',None,La pena máxima,None,89,La Pena Maxima
9,6530,2024-01-01,2024-01-01,b'\x00',None,Mirciulică,None,92,Mirciulica


### ¿Coinciden los resultados (cantidad de NULL y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.

Basado en el análisis realizado en las celdas anteriores, la canrtidad de peliculas donde no tiene un `locale` definido y su titulo original difiere de su titulo, **no coincide con la cuenta realizada de titulos con su `locale` nula**, ya que la cantidad de registros que satisfacen ambas clausulas es de `3947` y la cantidad de locales nulos es de `11255`.

Basado en el query anterior, podemos ver que la clausula de `=` no está tomando en cuenta puntuación ni acentos, un ejemplo de esto es la fila con `id` `1349` ya que identifica `Verónica` y `Veronica` como el mismo titulo, aunque el `original_title` contenga una tilde adicional.

Para este tipo de análisis, se debe tomar la decisión en si una película con las mismas letras pero diferente puntuación u acentuación, debería ser categorizada como una pelicula que mantuvo su titulo original en el mercado estadounidense.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [91]:
query_db("""
  SELECT m.title, MAX(vm.cumulative_weeks_in_top10) AS weeks_top10
  FROM movie m
  JOIN view_summary vm ON vm.movie_id = m.id
  WHERE vm.cumulative_weeks_in_top10 IS NOT NULL
  GROUP BY m.id, m.title
  ORDER BY weeks_top10 DESC
  LIMIT 1
""")

,title,weeks_top10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [92]:
query_db("""
  SELECT ts.id, ts.title, ts.original_title, count(s.id) as seasons_count
  FROM tv_show ts
  JOIN season s ON ts.id = s.tv_show_id
  GROUP BY ts.id, ts.title, ts.original_title
  ORDER BY seasons_count DESC
  LIMIT 5
""")

,id,title,original_title,seasons_count
0,347,Grey's Anatomy,None,20
1,1020,Naruto Shippuden,NARUTO-ナルト- 疾風伝,20
2,564,Heartland (2007),None,17
3,2207,It's Always Sunny in Philadelphia,None,16
4,881,Supernatural (2005),None,15


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [93]:
query_db("""
  SELECT DISTINCT start_date, end_date
  FROM view_summary
  WHERE duration = 'SEMI_ANNUALLY'
  ORDER BY start_date
""")

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [94]:
query_db("""
  SELECT s.id, s.title, vs.views
  FROM season s
  JOIN view_summary vs ON s.id = vs.season_id
  JOIN tv_show ts ON ts.id = s.tv_show_id
  WHERE vs.start_date = '2024-01-01'
    AND vs.end_date = '2024-06-30'
    AND vs.duration = 'SEMI_ANNUALLY'
    AND ts.title = 'Grey''s Anatomy'
  ORDER BY vs.views DESC
""")

,id,title,views
0,575,Grey's Anatomy: Season 1,3600000
1,685,Grey's Anatomy: Season 2,3100000
2,742,Grey's Anatomy: Season 3,2900000
3,743,Grey's Anatomy: Season 5,2900000
4,749,Grey's Anatomy: Season 4,2900000
5,776,Grey's Anatomy: Season 6,2800000
6,818,Grey's Anatomy: Season 7,2700000
7,850,Grey's Anatomy: Season 8,2600000
8,882,Grey's Anatomy: Season 9,2500000
9,915,Grey's Anatomy: Season 10,2400000


### ¿Cómo interpretarías los resultados obtenidos?

Durante el primer semestre de 2024, las vistas por temporada de Grey's Anatomy disminuyen casi en el mismo orden cronológico de la serie: de 3.6M en la temporada 1 a 1.7M en la 18, una retención aproximada del 47% (≈4% menos por temporada). Esto sugiere un embudo de abandono: muchos usuarios inician la serie, pero la extensión del catálogo (20 temporadas) reduce la proporción que avanza. Es probable que el estreno de la temporada 20 haya motivado a nuevos espectadores a comenzar desde el inicio. Destacan dos anomalías: la temporada 19 (2.0M) supera a las temporadas 15–18, posiblemente por usuarios que saltan a la temporada más reciente, y la temporada 20 registra solo 100K vistas, lo que apunta a una disponibilidad parcial dentro del periodo analizado.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [103]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [96]:
# Filtrar la serie
greys = tv_show.loc[tv_show['title'] == "Grey's Anatomy", ['id']]

# Filtrar el periodo semestral
vs = view_summary.copy()
vs['start_date'] = pd.to_datetime(vs['start_date'])
vs['end_date'] = pd.to_datetime(vs['end_date'])

vs_periodo = vs.loc[
    (vs['start_date'] == '2024-01-01') &
    (vs['end_date'] == '2024-06-30') &
    (vs['duration'] == 'SEMI_ANNUALLY'),
    ['season_id', 'views']
]

# JOINs: season a tv_show, luego season a view_summary
resultado = (
    season[['id', 'title', 'tv_show_id']]
    .merge(greys, left_on='tv_show_id', right_on='id', suffixes=('', '_show'))
    .merge(vs_periodo, left_on='id', right_on='season_id')
    [['id', 'title', 'views']]
    .sort_values('views', ascending=False)
    .reset_index(drop=True)
)

resultado

,id,title,views
0,575,Grey's Anatomy: Season 1,3600000
1,685,Grey's Anatomy: Season 2,3100000
2,742,Grey's Anatomy: Season 3,2900000
3,743,Grey's Anatomy: Season 5,2900000
4,749,Grey's Anatomy: Season 4,2900000
5,776,Grey's Anatomy: Season 6,2800000
6,818,Grey's Anatomy: Season 7,2700000
7,850,Grey's Anatomy: Season 8,2600000
8,882,Grey's Anatomy: Season 9,2500000
9,915,Grey's Anatomy: Season 10,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [97]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [98]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [99]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/"
os.chdir(DIR)

In [100]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [101]:
employees = pd.DataFrame(
    [{'id': doc.id, **doc.to_dict()} for doc in db.collection('EMPLOYEE').stream()]
)
employees

,id,emp_fname,emp_lname,emp_hiredate
0,8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
1,Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
2,lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
3,yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [102]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Anthropic. (2026). Claude (basado en Claude Opus 5.5) [Modelo de lenguaje grande], utilizado para depuración de consultas SQL y generación de código en Pandas. https://claude.ai/

---